aten/src/ATen/native/sparse/SparseMatMul.cpp

```diff
@@ -0,0 +1,231 @@
+#include <ATen/ATen.h>
+#include <ATen/Config.h>
+#include <ATen/NamedTensorUtils.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/Parallel.h>
+#include <ATen/SparseTensorImpl.h>
+#include <ATen/SparseTensorUtils.h>
+
+namespace at {
+namespace native {
+
+// helper
+namespace {
+
+template <typename scalar_t>
+std::vector<int64_t> sort_indices(const TensorAccessor<scalar_t, 1>& v) {
+  std::vector<int64_t> indices(v.size(0));
+  std::iota(indices.begin(), indices.end(), 0);
+  std::stable_sort(
+      indices.begin(),
+      indices.end(),
+      [&](const scalar_t& i1, const scalar_t& i2) { return v[i1] < v[i2]; });
+  return indices;
+}
+
+template <size_t N>
+std::vector<int64_t> indices2csr(
+    const TensorAccessor<int64_t, N>& indices,
+    int64_t dim) {
+  auto nnz = indices.size(0);
+  std::vector<int64_t> csr(dim + 1, int64_t(0));
+  int64_t last_i = 0;
+  for (int64_t index = 0; index < nnz; index++) {
+    int64_t i = indices[index];
+    if (i == last_i) {
+      csr[last_i + 1] += 1;
+    } else {
+      for (int64_t j = last_i; j < i + 1; j++) {
+        csr[j + 1] = csr[last_i + 1];
+      }
+      last_i = i;
+      csr[last_i + 1] += 1;
```

Comment:
This line can be pulled out of the conditional, right? If that happens the conditional reduces to a single branch, I think:

```
if (i != last_i) {
  for (int64_t j = last_i; j < i + 1; j++) {
     csr[j + 1] = csr[last_i + 1];
  }
last_i = i;
}
csr[last_i + 1] += 1;
```